patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch

```diff
@@ -397,6 +397,22 @@
        if (!this.f_20911_ || this.f_20913_ >= this.m_21304_() / 2 || this.f_20913_ < 0) {
           this.f_20913_ = -1;
           this.f_20911_ = true;
+@@ -1784,9 +_,12 @@
+    }
+ 
+    private void m_21312_() {
+-      ItemStack itemstack = this.m_6844_(EquipmentSlot.OFFHAND);
+-      this.m_8061_(EquipmentSlot.OFFHAND, this.m_6844_(EquipmentSlot.MAINHAND));
+-      this.m_8061_(EquipmentSlot.MAINHAND, itemstack);
++      final ItemStack offHand = this.m_6844_(EquipmentSlot.OFFHAND);
++      final ItemStack mainHand = this.m_6844_(EquipmentSlot.MAINHAND);
++      final var event = net.minecraftforge.common.ForgeHooks.onSwapHandItems(this, mainHand, offHand);
```

Comment:
Can’t the player just be passed to the hook instead of having these variables?

Reply:
They be fetched inside the event constructor, but they need to be there to be returned back from the event.

Reply:
I’m not sure if this is right. The code already calls event.getMainHand(), which means you can return whatever you want there, without regard to what local variables are present before the forge hook.

In fact the entire patch could be one line, since you can set the player’s hand items from anywhere.

Reply:
It'd need to be fetched then in the hook function or event constructor, to initialise the event fields. Doesn't really matter where the code is though, it's functionally the same either way.  With the current code it'd make it possible for a mod to manually fire the event with other items, though I'm not sure how useful that would be.

Reply:
You don't really need to pass the main hand and offhand with it, you can just pass the entity and get the hands from that, firing the event from a mod would be pointless and should generally not be done

But this event looks really promising, could add some functionality of having multiple items to swap from